### PR TITLE
ScoreX

### PR DESCRIPTION
**Pull request by gtklocker**
This PR is on behalf of IOHK. Responsible contacts: @ahmadalkabra, @romainPellerin.

# Short Description

ScoreX is a modular blockchain framework built with flexibility in mind to accelerate the design of experimental DLT technologies. ScoreX is the hyperflexible answer to the Fabric and Sawtooth frameworks. ScoreX is developed and maintained by IOHK and the founders of Ergo under the Apache 2.0 license.

### CODE REVIEW

**Review by vipinsun — COMMENTED** [review]
Please change DTL  to DLT. Sounds fascinating...

File: labs/scorex.md
Comment (tkuhrt) [inline labs/scorex.md:11]: [Hyperledger's charter](https://hyperledger.org/about/charter) has the following requirements regarding licensing:

> 12. Intellectual Property Policy
> 
> a. Members agree that all new inbound code contributions to HLP shall be made under the Apache License, Version 2.0 (available at http://www.apache.org/licenses/LICENSE-2.0). All contributions shall be accompanied by a Developer Certificate of Origin sign-off (http://developercertificate.org) that is submitted through a Governing Board and LF-approved contribution process. Such contribution process will include steps to also bind non-Member Contributors and, if not self-employed, their employer, to the licenses expressly granted in the Apache License, Version 2.0 with respect to such contribution.
> 
> b. All outbound code will be made available under the Apache License, Version 2.0.
> 
> c. All documentation will be contributed to and made available by HLP under the Creative Commons Attribution 4.0 International License (available at http://creativecommons.org/licenses/by/4.0/).
> 
> d. If an alternative inbound or outbound license is required for compliance with the license for a leveraged open source project or is otherwise required to achieve HLP’s mission, the Governing Board may approve the use of an alternative license for specific inbound or outbound contributions on an exception basis. Any exceptions must be approved by a two-thirds vote of the entire Governing Board and the LF and must be limited in scope to what is required for such purpose. Please email legal@hyperledger.org to obtain exception approval.
> 
> e. Subject to available Project funds, HLP may engage The Linux Foundation to determine the availability of, and register, trademarks, service marks, and certification marks, which shall be owned by the LF.

Comment (tkuhrt) [reply labs/scorex.md:11]: Two pieces of potential concern:
1. Source code license in Hyperledger is Apache 2.0. Are you willing to change the license?
2. Is ScoreX a trademark that is already owned by Ergo?

File: labs/scorex.md
Comment (tkuhrt) [inline labs/scorex.md:51]: Hyperledger requires that all contributed code has a DCO sign off. Looking at this repository, it appears that this is not the case. There are two options for fixing this:
1. Amend existing commits in the repository to include DCO sign off
2. Manually transfer the code from the existing repo to the newly created repo within the hyperledger-labs github organization.

File: labs/scorex.md
Comment (tkuhrt) [inline labs/scorex.md:47]: @hartm  : Can you confirm sponsorship?

File: labs/scorex.md
Comment (kushti) [reply labs/scorex.md:11]: @tkuhrt 
1) Yes, we can discuss license change
2) Trademark is not owned by any party

File: labs/scorex.md
Comment (kushti) [reply labs/scorex.md:51]: We can go either way likely, but need to discuss with colleagues

File: labs/scorex.md
Comment (hartm) [reply labs/scorex.md:47]: Yes!  Sorry for the late response.

File: labs/scorex.md
Comment (tkuhrt) [reply labs/scorex.md:11]: Let us know when you have discussed the license change and your decision, @kushti.

File: labs/scorex.md
Comment (vipinsun) [reply labs/scorex.md:47]: So as far I can see we have:
1. Small issue (DTL->DLT)
2. Apache 2.0 license for existing code
3. DCO signoff for existing code
Please confirm  @ahmadalkabra, @romainPellerin

File: labs/scorex.md
Comment (gtklocker) [reply labs/scorex.md:51]: @tkuhrt @kushti so as to avoid losing all the git history or altering old commits, I propose we get contributors to explicitly sign [the DCO](https://developercertificate.org/) for their existing commits in some place in the repo. This way we maintain the git history and don't need to alter old commits.

File: labs/scorex.md
Comment (gtklocker) [reply labs/scorex.md]: License change PR is up https://github.com/ScorexFoundation/Scorex/pull/385

File: labs/scorex.md
Comment (ahmadalkabra) [reply labs/scorex.md:51]: Sounds like a plan. @tkuhrt are you ok with this approach?

Comment (tkuhrt) [reply labs/scorex.md:51]: Unfortunately, all commits must have DCO signoff before we can bring over an existing repository in the lab. I have seen others who were able to add the DCO to the commits without modifying the history of the commits.

File: labs/scorex.md
Comment (tkuhrt) [reply labs/scorex.md]: I am going to mark this as unresolved until the PR is merged.

File: labs/scorex.md
Comment (gtklocker) [reply labs/scorex.md:51]: You are able to retroactively add DCO to every commit in history however the commit hashes will change and this may break compatibility with existing downstream users that depend on them. Since this is a project with real downstream users I would like to avoid any disruption. Additionally modifying the people's commits to make it seem as if they were singed off at the time seems more dishonest than them clearly agreeing *now* that their contributions then fell under the DCO.

Comment (gtklocker) [reply labs/scorex.md:51]: The Linux Foundation (writers of the DCO) also [prefers this approach](https://opensource.stackexchange.com/a/6722) for repos that did not start with signed-off commits.

Comment (gtklocker) [reply labs/scorex.md:51]: I pushed for the recommended approach in this PR: https://github.com/ScorexFoundation/Scorex/pull/390

Comment (tkuhrt) [reply labs/scorex.md:51]: The DCO signoff on all commits is a requirement of Hyperledger Charter. @ryjones : Can you comment on whether Hyperledger would accept this approach for DCO signoffs for a existing repo that we would transfer to Hyperledger Labs? It was my understanding that are options for DCO signoffs in Hyperledger is one of the following:
1) All commits have DCO signoff and we can bring existing repo over
-or-
2) We create a new repository and the code is brought over with a single commit that has DCO signoff

Comment (gtklocker) [reply labs/scorex.md:51]: @ryjones you feedback would be appreciated here :)

Comment (vipinsun) [reply labs/scorex.md:51]: I for one am in favor of reducing barriers to entry. There have to be multiple paths to cure DCO from existing code. We have to think about why DCO is required.
Mostly to handle IP issues, with developers signing off that they have the right to contribute the code that they have checked in and it is free of IP issues. However, this can fail for many reasons including bad faith by the developer who is signing off code that they do not have a right to.
Given the nature of this DCO signoff it is good to have alternate pathways to enforce what is at best an imperfect solution, including a txt file with various people signing off on hashes that they are responsible for checked into the docs folder. I believe the LF has to opine on this, but putting barriers across valuable contributions through strict adherence to the letter and not the spirit of the text is not the way to do this. @gtklocker @tkuhrt  @ryjones @lehors @trbs

Comment (lehors) [reply labs/scorex.md:51]: I agree with Vipin and I think it can be argued that having a text file that lists all the commits by associating their hashes with the relevant signoff still fulfills the requirement. Indeed, "All commits have DCO signoff" doesn't say that the only way this can be done is in the commit log, does it?
Don't get me wrong, I don't think that's ideal and wouldn't want this to continue for future commits but I do think it would fulfill the requirement of ensuring there is a legal commitment for every commit.
For what it's worth, I think this gives us more than a single commit against a new repo which squashes history entirely.

Comment (ryjones) [reply labs/scorex.md:51]: Sorry, I missed this. We’ve agreed that you can have a commit that lists the commit ids of your commits and make a DCO statement in that commit. —allow-empty

Comment (vipinsun) [reply labs/scorex.md:51]: So I believe the ball is in your court @gtklocker @kushti. Or am I wrong?

Comment (gtklocker) [reply labs/scorex.md:51]: Yes, thank you. @kushti will take it from here. The process of retroactively applying the DCO has started in this PR: https://github.com/ScorexFoundation/Scorex/pull/390

File: labs/scorex.md
Comment (kushti) [reply labs/scorex.md]: @tkuhrt merged!

File: labs/scorex.md
Comment (kushti) [reply labs/scorex.md:51]: @vipinsun @gtklocker @ryjones @lehors @tkuhrt merged!

**Comment by ahmadalkabra** [comment]
Hi @trbs @lehors can you please update your review. Thanks

File: labs/scorex.md
Comment (ahmadalkabra) [reply labs/scorex.md:51]: @vipinsun @ryjones @lehors all the issues have been resolved.  Can you please update the review and let us know the next steps.  Thanks!

File: labs/scorex.md
Comment (lehors) [inline labs/scorex.md:11]: "ScoreX is the hyperflexible answer to the Fabric and Sawtooth frameworks." seems unnecessarily adversarial and at odd with the much better and more complete positioning that is done below with regard to all the different DLT frameworks we currently have at Hyperledger.

**Review by lehors — CHANGES_REQUESTED** [review]

File: labs/scorex.md
Comment (ahmadalkabra) [reply labs/scorex.md:11]: It should probably say "complementary" which is the intention.

**Review by lehors — APPROVED** [review]

File: labs/scorex.md
Comment (lehors) [reply labs/scorex.md]: thanks

**Comment by ahmadalkabra** [comment]
@trbs can you please review so we can move this forward to the next steps.  thanks!

**Comment by trbs** [comment]
lgtm

@hartm could you confirm sponsorship ?
@tkuhrt @lehors the DCO is solved by ScorexFoundation/Scorex#390 right ?

**Comment by hartm** [comment]
> @hartm could you confirm sponsorship ?

I think I already did, but am happy to do so again.  Thanks!

**Review by trbs — APPROVED** [review]

File: labs/scorex.md
Comment (ryjones) [reply labs/scorex.md:51]: @ahmadalkabra please transfer the existing repo, https://github.com/ScorexFoundation/Scorex, to me.

Comment (ahmadalkabra) [reply labs/scorex.md:51]: @kushti can you help with this request

Comment (kushti) [reply labs/scorex.md:51]: @ryjones done! "Repository transfer to ryjones requested" (@ahmadalkabra)

**Comment by ryjones** [comment]
https://github.com/hyperledger-labs/Scorex

**Pull request merged**
